# Exercise 4.6 Combining and exporting data 

### The scripts contain the following steps:
#### 1. Importing librairies and dataframe
#### 2. Dropping a column in the products dataframe
#### 3. Combining df_prods and orders_products_combined
#### 4. Exporting the ords_prods_merge dataframe

#### Importing libraries and dataframe 

In [64]:
# Importing librairies
import pandas as pd
import numpy as np
import os

In [65]:
path = r'C:\Users\salmi\Desktop\Projects\Achievement 4\Instacart Basket Analysis'

In [66]:
#Importing dataframe orders_products_prior
df_orders_products_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

In [67]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_clean.csv'), index_col = False)

In [68]:
df_orders_products_combined.shape

(32434489, 10)

In [69]:
df_prods

Unnamed: 0  product_id  \
0               0           1   
1               1           2   
2               2           3   
3               3           4   
4               4           5   
...           ...         ...   
49667       49688       49684   
49668       49689       49685   
49669       49690       49686   
49670       49691       49687   
49671       49692       49688   

                                            product_name  aisle_id  \
0                             Chocolate Sandwich Cookies        61   
1                                       All-Seasons Salt       104   
2                   Robust Golden Unsweetened Oolong Tea        94   
3      Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4                              Green Chile Anytime Sauce         5   
...                                                  ...       ...   
49667          Vodka, Triple Distilled, Twist of Vanilla       124   
49668                 En Croute Roast Hazelnut Cranberry        42   
49669                                   Artisan Baguette       112   
49670         Smartblend Healthy Metabolism Dry Cat Food        41   
49671                             Fresh Foaming Cleanser        73   

       department_id  prices  
0                 19     5.8  
1                 13     9.3  
2                  7     4.5  
3                  1    10.5  
4                 13     4.3  
...              ...     ...  
49667              5     5.3  
49668              1     3.1  
49669              3     7.8  
49670              8     4.7  
49671             11    13.5  

[49672 rows x 6 columns]

#### 2. Dropping a column in the products dataframe

In [71]:
# deleting 1st column prods
df_prods = df_prods.drop(columns=['Unnamed: 0'])

In [72]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49667       49684          Vodka, Triple Distilled, Twist of Vanilla   
49668       49685                 En Croute Roast Hazelnut Cranberry   
49669       49686                                   Artisan Baguette   
49670       49687         Smartblend Healthy Metabolism Dry Cat Food   
49671       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49667       124              5     5.3  
49668        42              1     3.1  
49669       112              3     7.8  
49670        41              8     4.7  
49671        73             11    13.5  

[49672 rows x 5 columns]

In [73]:
df_orders_products_combined

order_id  user_id  order_amount  orders_day_of_week  \
0          2539329        1             1                   2   
1          2539329        1             1                   2   
2          2539329        1             1                   2   
3          2539329        1             1                   2   
4          2539329        1             1                   2   
...            ...      ...           ...                 ...   
32434484   2977660   206209            13                   1   
32434485   2977660   206209            13                   1   
32434486   2977660   206209            13                   1   
32434487   2977660   206209            13                   1   
32434488   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         8                     NaN       14084   
2                         8                     NaN       12427   
3                         8                     NaN       26088   
4                         8                     NaN       26405   
...                     ...                     ...         ...   
32434484                 12                     7.0       14197   
32434485                 12                     7.0       38730   
32434486                 12                     7.0       31477   
32434487                 12                     7.0        6567   
32434488                 12                     7.0       22920   

          add_to_cart_order  reordered _merge  
0                         1          0   both  
1                         2          0   both  
2                         3          0   both  
3                         4          0   both  
4                         5          0   both  
...                     ...        ...    ...  
32434484                  5          1   both  
32434485                  6          0   both  
32434486                  7          0   both  
32434487                  8          0   both  
32434488                  9          0   both  

[32434489 rows x 10 columns]

### /!\ As the data set contains the columns _merge, I will delete it to be able to have a new merged data set with a new _merge column for only these 2 specific datasets

In [75]:
# deleting _merge column from df_orders_products_combined
df_orders_products_combined = df_orders_products_combined.drop(columns=['_merge'])

In [104]:
df_orders_products_combined.shape

(32434489, 9)

### 3. Combining df_prods and orders_products_combined

In [100]:
df_ords_prods_merge = df_orders_products_combined.merge(df_prods, on = 'product_id', indicator = True)

In [101]:
df_prods.shape

(49672, 5)

In [78]:
df_ords_prods_merge.head()

order_id  user_id  order_amount  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

In [79]:
df_ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

#### (I also did a check with an how:outer to check, that had the following result
_merge
both 32404859
right_only 30200
left_only 1)

#### 4. Exporting the ords_prods_merge dataframe

In [82]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge.pkl'))

In [106]:
# Check for missing values in the 'prices' column
missing_values_in_prices = df_ords_prods_merge['prices'].isnull().sum()

In [108]:
missing_values_in_prices

0

In [112]:
missing_values_in_prices_2 = df_ords_prods_merge.isnull().sum()

In [114]:
missing_values_in_prices_2

order_id                        0
user_id                         0
order_amount                    0
orders_day_of_week              0
order_hour_of_day               0
days_since_prior_order    2076096
product_id                      0
add_to_cart_order               0
reordered                       0
product_name                    0
aisle_id                        0
department_id                   0
prices                          0
_merge                          0
dtype: int64